nexus/db-queries/src/db/datastore/affinity.rs

```diff
@@ -0,0 +1,2701 @@
+            .execute_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+        Ok(())
+    }
+
+    pub async fn affinity_group_member_delete(
+        &self,
+        opctx: &OpContext,
+        authz_affinity_group: &authz::AffinityGroup,
+        member: external::AffinityGroupMember,
+    ) -> Result<(), Error> {
+        opctx.authorize(authz::Action::Modify, authz_affinity_group).await?;
+
+        let instance_id = match member {
+            external::AffinityGroupMember::Instance(id) => id,
+        };
+
+        let err = OptionalError::new();
+        let conn = self.pool_connection_authorized(opctx).await?;
+        self.transaction_retry_wrapper("affinity_group_member_delete")
+            .transaction(&conn, |conn| {
+                let err = err.clone();
+                use db::schema::affinity_group::dsl as group_dsl;
+                use db::schema::affinity_group_instance_membership::dsl as membership_dsl;
+                use db::schema::instance::dsl as instance_dsl;
+
+                async move {
+                    // Check that the group exists
+                    group_dsl::affinity_group
+                        .filter(group_dsl::time_deleted.is_null())
+                        .filter(group_dsl::id.eq(authz_affinity_group.id()))
+                        .select(group_dsl::id)
+                        .first_async::<uuid::Uuid>(&conn)
+                        .await
+                        .map_err(|e| {
+                            err.bail_retryable_or_else(e, |e| {
+                                public_error_from_diesel(
+                                    e,
+                                    ErrorHandler::NotFoundByResource(
+                                        authz_affinity_group,
+                                    ),
+                                )
+                            })
+                        })?;
+
+                    // Check that the instance exists
+                    instance_dsl::instance
+                        .filter(instance_dsl::time_deleted.is_null())
+                        .filter(instance_dsl::id.eq(instance_id))
+                        .select(instance_dsl::id)
+                        .first_async::<uuid::Uuid>(&conn)
+                        .await
+                        .map_err(|e| {
+                            err.bail_retryable_or_else(e, |e| {
+                                public_error_from_diesel(
+                                    e,
+                                    ErrorHandler::NotFoundByLookup(
+                                        ResourceType::Instance,
+                                        LookupType::ById(instance_id)
+                                    ),
+                                )
+                            })
+                        })?;
```

Comment:
Hmm, if we removed this check, it seems like we wouldn't need to do this in a transaction. If we don't find any rows matching, that _implies_ that either the group or the instance don't exist, right? We could just do the `delete` outside of a transaction, and if no rows were deleted, we could then go check whether the instance or affinity group are nonexistent in order to decide what error message to return.

Unlike the `(anti_)affinity_group_member_add` queries, we won't do something bad if we don't first check that the instance and group exist, we'll just not delete anything, which doesn't cause any harm...

Reply:
I *think* I still need to do this in a transaction to get a better error out - I need to identify if "the group is missing, or if the membership is missing" - but you're totally right, there's no need to check the instance here. I'll remove that.

Reply:
(Patched in 1ad01010bcd8e59581a267463b3efb671a748343 )